crates/matrix-sdk-ui/src/sync_service.rs

```diff
@@ -88,41 +84,38 @@ impl SyncTaskSupervisor {
         room_list_service: Arc<RoomListService>,
         encryption_sync_permit: Arc<AsyncMutex<EncryptionSyncPermit>>,
     ) -> Self {
-        let (sender, receiver) = tokio::sync::mpsc::channel(16);
-        let (room_list_task, encryption_sync_task) = Self::spawn_child_tasks(
-            inner,
-            room_list_service.clone(),
-            encryption_sync_permit,
-            sender.clone(),
-        )
-        .await;
+        let (task, termination_sender) =
+            Self::spawn_supervisor_task(inner, room_list_service, encryption_sync_permit).await;
 
-        let task = spawn(Self::spawn_supervisor_task(
-            inner,
-            room_list_service,
-            room_list_task,
-            encryption_sync_task,
-            receiver,
-        ));
-
-        Self { task, termination_sender: sender }
+        Self { task, termination_sender }
     }
 
     /// The role of the supervisor task is to wait for a termination message
     /// ([`TerminationReport`]), sent either because we wanted to stop both
     /// syncs, or because one of the syncs failed (in which case we'll stop the
     /// other one too).
-    fn spawn_supervisor_task(
+    async fn spawn_supervisor_task(
         inner: &SyncServiceInner,
         room_list_service: Arc<RoomListService>,
-        room_list_task: JoinHandle<()>,
-        encryption_sync_task: JoinHandle<()>,
-        mut receiver: Receiver<TerminationReport>,
-    ) -> impl Future<Output = ()> {
+        encryption_sync_permit: Arc<AsyncMutex<EncryptionSyncPermit>>,
+    ) -> (JoinHandle<()>, Sender<TerminationReport>) {
+        let (sender, mut receiver) = tokio::sync::mpsc::channel(16);
+
         let encryption_sync = inner.encryption_sync_service.clone();
         let state = inner.state.clone();
+        let termination_sender = sender.clone();
+
+        let sync_permit_guard = encryption_sync_permit.clone().lock_owned().await;
+
+        let future = async move {
+            let (room_list_task, encryption_sync_task) = Self::spawn_child_tasks(
+                room_list_service.clone(),
+                encryption_sync.clone(),
+                sync_permit_guard,
+                sender.clone(),
```

Comment:
This `clone()` might be spurious, since it looks like it's the last use of `sender`.

Reply:
It's used inside of a loop so it's necessary, though you might be looking at a commit before the loop was introduced.